crates/consensus-logic/src/fork_choice_manager.rs

```diff
@@ -289,19 +301,48 @@ fn forkchoice_manager_task_inner<D: Database, E: ExecEngineCtl>(
     engine: &E,
     mut fcm_rx: mpsc::Receiver<ForkChoiceMessage>,
     csm_ctl: &CsmController,
+    mut cupdate_rx: Receiver<Arc<ClientUpdateNotif>>,
 ) -> anyhow::Result<()> {
-    loop {
+    let mut notif_closed = false;
+    let mut fcm_closed = false;
+
+    while !notif_closed || !fcm_closed {
         if shutdown.should_shutdown() {
             warn!("received shutdown signal");
             break;
         }
+        match cupdate_rx.try_recv() {
+            Ok(update_notif) => {
+                handle_new_state(
+                    &mut state,
+                    update_notif.new_state(),
+                    update_notif.tsn_output(),
+                )?;
+            }
+            Err(broadcast::error::TryRecvError::Empty) => {}
+            Err(broadcast::error::TryRecvError::Closed) => {
+                notif_closed = true;
+            }
+            Err(broadcast::error::TryRecvError::Lagged(_)) => {
+                // TODO: what to do?
+                notif_closed = true;
+            }
+        }
 
-        let Some(m) = fcm_rx.blocking_recv() else {
-            break;
-        };
+        match fcm_rx.try_recv() {
+            Ok(m) => {
+                // TODO decide when errors are actually failures vs when they're okay
+                process_fc_message(m, &mut state, engine, csm_ctl)?;
+            }
+            Err(mpsc::error::TryRecvError::Empty) => {}
+            Err(mpsc::error::TryRecvError::Disconnected) => {
+                // mark done
+                fcm_closed = true;
+            }
+        }
 
-        // TODO decide when errors are actually failures vs when they're okay
-        process_fc_message(m, &mut state, engine, csm_ctl)?;
+        // Just to avoid the tight loop
+        std::thread::sleep(Duration::from_millis(100));
```

Comment:
Remove this or make this configurable.
100ms / message adds a very hard bound to throughput.

Reply:
Removing it altogether with the use of watch channel.